sympy/matrices/expressions/diagonal.py

```diff
@@ -5,14 +5,14 @@
 
 class DiagonalMatrix(MatrixExpr):
     arg = property(lambda self: self.args[0])
-    shape = property(lambda self: (self.arg.shape[0], self.arg.shape[0]))
+    shape = property(lambda self: (self.arg.shape[0], self.arg.shape[1]))
 
     def _entry(self, i, j):
-        return S.Zero if i != j else self.arg[i, 0]
+        return S.Zero if i != j else self.arg[i, i]
```

Comment:
This is better but still wrong in that it gives 0 for `self[i,j]` when `i,j` is unknown. Someone else is working on this.